File: scan.py
"""
Scan Router — /api/scan
Accepts an email payload, runs it through the 8-layer engine,
returns a full EAA threat report.
"""
from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, EmailStr
from typing import Optional
import structlog

from engine.layer1_envelope import analyze_envelope
from engine.layer2_reputation import analyze_reputation
from engine.layer3_headers import analyze_headers
from engine.layer4_linguistic import analyze_linguistics
from engine.layer5_impersonation import analyze_impersonation
from engine.layer6_links import analyze_links
from engine.layer7_behavioral import analyze_behavioral
from engine.layer8_composite import compute_composite
from models.scan import ScanRecord
from config import get_settings

logger = structlog.get_logger()
router = APIRouter()
settings = get_settings()


# ── Request / Response Models ──────────────────────────────────
class EmailScanRequest(BaseModel):
    """
    Raw email data submitted for scanning.
    Data contract — must match frontend payload exactly.
    """
    raw_headers:   str
    from_address:  str
    from_name:     Optional[str] = None
    reply_to:      Optional[str] = None
    subject:       str
    body_text:     str
    body_html:     Optional[str] = None
    attachments:   list[str] = []       # List of attachment filenames
    user_id:       str                  # Supabase user UUID


class LayerResult(BaseModel):
    layer:       int
    name:        str
    score:       float                  # 0-100
    multiplier:  float
    weighted:    float
    signals:     list[str]             # Human-readable EAA explanations
    veto:        bool = False


class ScanResponse(BaseModel):
    scan_id:         str
    composite_score: float
    classification:  str               # SAFE / SUSPICIOUS / QUARANTINE
    veto_triggered:  bool
    action:          str               # deliver / warn / quarantine
    layers:          list[LayerResult]
    summary:         str               # One-sentence plain-language verdict


# ── Endpoint ───────────────────────────────────────────────────
@router.post("/", response_model=ScanResponse)
async def scan_email(payload: EmailScanRequest):
    """
    Run an email through all 8 layers and return the full EAA threat report.
    This is the core pipeline. All layers are run in sequence.
    Layer 8 computes the weighted composite and applies veto logic.
    """
    log = logger.bind(user_id=payload.user_id, from_address=payload.from_address)
    log.info("scan_started")

    try:
        # ── Run all 7 detection layers ─────────────────────────
        l1 = await analyze_envelope(payload)
        l2 = await analyze_reputation(payload)
        l3 = await analyze_headers(payload)
        l4 = await analyze_linguistics(payload)
        l5 = await analyze_impersonation(payload)
        l6 = await analyze_links(payload)
        l7 = await analyze_behavioral(payload)

        layer_results = [l1, l2, l3, l4, l5, l6, l7]

        # ── Layer 8: Composite + EAA output ────────────────────
        result = compute_composite(layer_results)

        log.info(
            "scan_complete",
            composite=result["composite_score"],
            classification=result["classification"],
            veto=result["veto_triggered"],
        )

        return ScanResponse(**result)

    except Exception as e:
        log.error("scan_failed", error=str(e))
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Scan pipeline failed. Please retry.",
        )

File: frontend/scan.ts
/**
 * TalonVigil — shared scan types.
 * These MIRROR the backend contract exactly (backend/api/routers/scan.py).
 * Keep them in sync with the Pydantic models.
 */

export type Classification = "SAFE" | "SUSPICIOUS" | "QUARANTINE" | "VETO";

/** One detection layer's output (backend LayerResult). Layers 1-7; L8 is the gate. */
export interface LayerResult {
  layer: number;
  name: string;
  score: number;        // 0-100 raw score
  multiplier: number;   // locked weight (L5 highest at 2.0)
  weighted: number;     // score * multiplier
  signals: string[];    // human-readable EAA explanations
  veto: boolean;        // this layer triggered a veto override
}

/** Full EAA threat report (backend ScanResponse). */
export interface ScanResult {
  scan_id: string;
  composite_score: number;
  classification: Classification;
  veto_triggered: boolean;
  action: string;                   // deliver / warn / quarantine
  layers: LayerResult[];
  summary: string;
}

/** Payload accepted by POST /api/scan/ (backend EmailScanRequest). user_id is server-populated. */
export interface EmailScanRequest {
  raw_headers: string;
  from_address: string;
  from_name?: string;
  reply_to?: string;
  subject: string;
  body_text: string;
  body_html?: string;
  attachments?: string[];
  user_id?: string;
}

/** A scan as it appears in history (mirrors a scan_records row). */
export interface ScanHistoryItem extends ScanResult {
  received_at: string;   // ISO timestamp
  from_address: string;
  subject: string;
}

/** Locked scoring constants — mirror backend/utils/scoring.py. */
export const QUARANTINE_THRESHOLD = 66;
export const SUSPICIOUS_THRESHOLD = 31;
export const VETO_THRESHOLD = 80;
export const VETO_ELIGIBLE_LAYERS = new Set([1, 5, 6]);
export const LAYER_WEIGHTS: Record<number, number> = {
  1: 1.8, 2: 1.3, 3: 1.2, 4: 1.4, 5: 2.0, 6: 1.9, 7: 1.1,
};
export const LAYER_NAMES: Record<number, string> = {
  1: "Envelope", 2: "Reputation", 3: "Headers",
  4: "Linguistics", 5: "Impersonation", 6: "Links", 7: "Behavioral",
};
